dali/operators/imgcodec/image_decoder.h

```diff
@@ -732,125 +681,174 @@ class ImageDecoder : public StatelessOperator<Backend> {
-        int block_idx = 0;
-        for (; block_idx < tp_->NumThreads(); block_idx++) {
-          tp_->AddWork(get_task(block_idx, nblocks), -block_idx);
+    if (ntasks < 2) {
+      setup_block(0, 1, -1);  // run all in current thread
+    } else {
+      int block_idx = 0;
+      atomic_idx_.store(0);
+      auto setup_task = [&, nblocks](int tid) {
+        DomainTimeRange tr("Setup", DomainTimeRange::kOrange);
+        int block_idx;
+        while ((block_idx = atomic_idx_.fetch_add(1)) < nblocks) {
+          setup_block(block_idx, nblocks, tid);
         }
-        tp_->RunAll(false);                 // start work but not wait
-        get_task(block_idx, nblocks)(-1);   // run last block
-        tp_->WaitForWork();                 // wait for the other threads
-      } else {                              // not worth parallelizing
-        get_task(0, 1)(-1);                 // run all in current thread
+      };
+
+      for (int task_idx = 0; task_idx < ntasks - 1; task_idx++) {
+        tp_->AddWork(setup_task, -task_idx);
       }
+      assert(ntasks >= 2);
+      tp_->RunAll(false);  // start work but not wait
+      setup_task(-1);      // last task in current thread
+      tp_->WaitForWork();  // wait for the other threads
+    }
 
-      for (int orig_idx : decode_sample_idxs_) {
-        auto &st = *state_[orig_idx];
+    bool any_need_processing = false;
+    bool has_any_roi = false;
+    for (int orig_idx = 0; orig_idx < nsamples; orig_idx++) {
+      auto &st = *state_[orig_idx];
+      any_need_processing |= state_[orig_idx]->need_processing;
+      if (use_cache && st.load_from_cache) {
+        auto *data_ptr = output.raw_mutable_tensor(orig_idx);
+        auto src_info = input.GetMeta(orig_idx).GetSourceInfo();
+        cache_->DeferCacheLoad(src_info, static_cast<uint8_t *>(data_ptr));
+      } else {
+        has_any_roi |= rois_[orig_idx].use_roi();
         batch_encoded_streams_.push_back(st.parsed_sample.encoded_stream);
         batch_images_.push_back(st.image);
+        decode_sample_idxs_.push_back(orig_idx);
       }
     }
+    size_t nsamples_decode = batch_images_.size();
+    size_t nsamples_cache = nsamples - nsamples_decode;
 
-    // This is a workaround for nvImageCodec <= 0.2
-    auto any_need_processing = [&]() {
-      for (int orig_idx : decode_sample_idxs_) {
-        auto& st = state_[orig_idx];
-        assert(ws.stream() == st->image_info.cuda_stream);  // assuming this is true
-        if (st->need_processing)
-          return true;
-      }
-      return false;
-    };
-    if (ws.has_stream() && need_host_sync_alloc() && any_need_processing()) {
+    if (ws.has_stream() && need_host_sync_alloc() && any_need_processing) {
       DomainTimeRange tr("alloc sync", DomainTimeRange::kOrange);
       CUDA_CALL(cudaStreamSynchronize(ws.stream()));
     }
 
-    {
-      DomainTimeRange tr("Decode", DomainTimeRange::kOrange);
+    if (nsamples_cache > 0) {
```

Comment:
```suggestion
    if (use_cache && nsamples_cache > 0) {
```
coverity detected it.
(don't ask me why there's coverity scan for this PR)